### PR TITLE
[MM-45564] - Notify Admin v2

### PR DESCRIPTION
**Pull request by AGMETEOR**
#### Summary
v2 notify admin

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-45564

#### Related Pull Requests

- Has webapp changes (https://github.com/mattermost/mattermost-webapp/pull/10892)

#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Comment by AGMETEOR** [comment]
I have added a do not merge label because I made a temporary change to make the job that pushes the notifications to the admins run faster.

File: api4/cloud.go
Comment (neallred) [inline api4/cloud.go:55]: Seems like this trigger endpoint is only needed to allow end to end tests to request the notifications to happen immediately instead of through the job. I see in the handler its locked down to admins, which is good.

Can we condition the adding of this endpoint on whether `config.EnableTesting` is set to true? I'm *guessing* e2e runs have that set to true (spinwicks definitely do; that's why the `/test threaded_post` commands work there.)

**Review by neallred — CHANGES_REQUESTED** [review]
Good stuff! Cool to see the new version of this with grouping. Looks like we can limit additions to the App struct, and there's a few other changes I think are worth making. Overall this looks like it will function as desired.

File: app/server.go
Comment (neallred) [reply app/server.go:1441]: Nevermind; just saw the comment that this is what it is for :+1:

File: api4/cloud.go
Comment (AGMETEOR) [reply api4/cloud.go:76]: Yes., this is just for testing.

File: app/cloud.go
Comment (AGMETEOR) [reply app/cloud.go:41]: The way it's implemented now, a user can notify once within the cool off period about a feature. Remember, we don't delete any notifications until they have been sent. Sending notifications only happens once within the cool off period.

File: db/migrations/postgres/000093_notify_admin.up.sql
Comment (AGMETEOR) [reply db/migrations/postgres/000093_notify_admin.up.sql:8]: Yes this is awesome. I am wondering (1/5) if we can first hold off on this because maybe requirements could change and duplicate requests could be allowed in the future (we won't have to do an alter migration). Atm, other teams have it that the notification keeps going out as long as the user keeps clicking notify.

File: app/server.go
Comment (AGMETEOR) [reply app/server.go:1444]: Oh yeah. I see they have made adding jobs this way much simpler than how it was before.

File: app/cloud.go
Comment (neallred) [reply app/cloud.go:40]: Oops, got confused with the job running every minute for testing and not realizing it only sends the notification once every 30 days no matter what. Looks like there's nothing to do here since we want them to be able to notify after the cool off. Thanks!

**Comment by neallred** [comment]
Looking much better! Could you keep looking into internationalization? If we switch to using ids that map to translation ids, that might help (and also make it so the ids the webapp passes aren't tied to mattermost-server internationalization internals). Looks like the makefile nit is pending. For the suggestion to conditionally enable the endpoint only if the config is a certain way, I think I found a better way to do it, reading some of the other existing code. The config defines API specific flags, for example [`EnableAPITeamDeletion`](https://github.com/mattermost/mattermost-server/tree/master/model/config.go#L753-755
). We can define `EnableAPITriggerNotifyAdmin` and use it in the handler, similar to how [`deleteTeam`](https://github.com/mattermost/mattermost-server/tree/master/api4/team.go#L440) makes use of us `EnableAPITeamDeletion`. Because that is a check each time the handler runs, it means e2e tests can enable it as needed. There's an example of an e2e test doing that here: https://github.com/mattermost/mattermost-webapp/e2e/cypress/tests/integration/team_settings/invite_members_spec.js:#L18-24

File: app/cloud.go
Comment (AGMETEOR) [inline app/cloud.go:258]: @neallred I feel like there's a better way to keep the translation ids in the en.json file. But not sure how. Could this suffice?

**Review by neallred — APPROVED** [review]
Server side LGTM, thanks @AGMETEOR !

File: app/cloud.go
Comment (marianunez) [inline app/cloud.go:131]: Should we also check if they are on trial? If they already on Enterprise trial, we should notify on that

File: app/cloud.go
Comment (marianunez) [inline app/cloud.go:71]: This is good for now, but thinking on what is coming to expand this to self-managed, wondering if we should check here for license sku instead of subscription sku. This will disassociate this feature from cloud and we can move this out of this `/cloud.go` 

(Not a blocker for this PR but can be done here if we don't want to change it later)

File: app/cloud.go
Comment (marianunez) [inline app/cloud.go:144]: Left a comment in Figma to see if we can drop the workspace name so we don't have to be cloud specific

File: jobs/notify_admin/scheduler.go
Comment (marianunez) [inline jobs/notify_admin/scheduler.go:15]: Is this intended to be 1 minute?

File: model/cloud.go
Comment (marianunez) [inline model/cloud.go:67]: We could use this license skus to be generic that are the same for cloud/self-managed under `shortNameSku` https://github.com/mattermost/mattermost-server/blob/7deb408898aaf88aa4644652d7f250a26fb5bdbc/model/license.go#L24-L25

**Review by marianunez — COMMENTED** [review]
Excellent work @AGMETEOR! Great attention to tests here and enabling for E2E as well 💯 

Only blocker would be to check the trial before sending - left some additional comments to make it not specific to cloud

File: app/cloud.go
Comment (AGMETEOR) [reply app/cloud.go:188]: My view is we should avoid duplicates. The first notification (if any), already serves the purpose the extra notification would.

File: app/cloud.go
Comment (AGMETEOR) [reply app/cloud.go:131]: Being on trial means the current SKU would be `enteprise`. I think it makes sense not to send notifications to trial a plan they're already trialing

File: app/cloud.go
Comment (AGMETEOR) [reply app/cloud.go:71]: Agreed

File: jobs/notify_admin/scheduler.go
Comment (AGMETEOR) [reply jobs/notify_admin/scheduler.go:15]: Yes, it's a temp change to make testing on the spinwicks much faster. There's a do not merge label to remind me to correct this to the right time

File: app/notify_admin.go
Comment (marianunez) [inline app/notify_admin.go:22]: Reminder to revert before merging

File: app/cloud.go
Comment (marianunez) [reply app/cloud.go]: Yes, but how can we be certain there was a first notification? We don't have the value to know because of the error so it may have not have been notified before and defaulting here to `true` then the request won't be sent and we would lose it

File: app/cloud.go
Comment (marianunez) [reply app/cloud.go]: Correct, I meant we shouldn't notify if they are already on trial. Good point that if they are on trial they are already in Enterprise 👍

File: app/notify_admin.go
Comment (marianunez) [inline app/notify_admin.go:54]: @AGMETEOR how is the text showing up here with the empty workspace name?

File: app/cloud.go
Comment (AGMETEOR) [reply app/cloud.go]: Alright. I have updated the code to default to `false`

File: app/notify_admin.go
Comment (AGMETEOR) [reply app/notify_admin.go:22]: Yes, the do not merge label is serving as my reminder

**Review by marianunez — APPROVED** [review]
Great stuff @AGMETEOR! 🎉 We just need to get a final call for the workspace name in the notification

File: app/notify_admin.go
Comment (AGMETEOR) [reply app/notify_admin.go:54]: It reads like this "10 members of **the workspace** have requested a workspace upgrade for:"

Comment (AGMETEOR) [reply app/notify_admin.go:54]: I should probably get rid of it since we shall be having this feature on self-hosted as well

**Review by yasserfaraazkhan — APPROVED** [review]

File: app/notify_admin.go
Comment (AGMETEOR) [reply app/notify_admin.go:54]: Hi @marianunez where did you post the suggestion to get rid of the workspace name so that I follow up?

**Pull request merged**
